functionaltests/main_test.go

```diff
@@ -0,0 +1,124 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package functionaltests
+
+import (
+	"context"
+	"flag"
+	"net/url"
+	"testing"
+
+	"github.com/elastic/apm-perf/pkg/telemetrygen"
+	"github.com/elastic/apm-server/functionaltests/internal/esclient"
+	"github.com/elastic/go-elasticsearch/v8/typedapi/types"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest"
+)
+
+var cleanupOnFailure *bool = flag.Bool("cleanup-on-failure", true, "Whether to run cleanup even if the test failed.")
+var target *string = flag.String("target", "production", "The target environment where to run tests againts. Valid values are: qa, production")
+
+// https://www.elastic.co/guide/en/cloud/current/ec-regions-templates-instances.html
+const testRegionQA = "aws-eu-west-1"
+const testRegionProduction = "eu-west-1"
+
+// assertDocCountEqual asserts that document counts in each data stream are equal.
+func assertDocCountEqual(t *testing.T, want []esclient.ApmDocCount, actual []esclient.ApmDocCount) {
+	t.Helper()
+
+	assert.Len(t, want, len(actual))
+	for i, v := range actual {
+		assert.Equal(t, v.Count, want[i].Count, "expected doc count in %s to be equal to previous value", v.Datastream)
+	}
+}
+
+// assertDocCountGreaterThan verifies if document count in datastreams is greated than expected.
+func assertDocCountGreaterThan(t *testing.T, want []esclient.ApmDocCount, actual []esclient.ApmDocCount) {
+	t.Helper()
+
+	assert.Len(t, want, len(actual))
+	for i, v := range actual {
+		assert.Greater(t, v.Count, want[i].Count, "expected doc count in %s to be greater than previous value", v.Datastream)
+	}
+}
+
+// ingest creates and run a telemetrygen that replays multiple APM agents events to the cluster
+// a single time.
+func ingest(t *testing.T, apmURL string, apikey string) {
+	t.Helper()
+
+	t.Log("ingest data")
+	cfg := telemetrygen.DefaultConfig()
+	cfg.APIKey = apikey
+
+	u, err := url.Parse(apmURL)
+	require.NoError(t, err)
+	cfg.ServerURL = u
+
+	cfg.EventRate.Set("1000/s")
+	g, err := telemetrygen.New(cfg)
+	g.Logger = zaptest.NewLogger(t, zaptest.Level(zap.InfoLevel))
+	require.NoError(t, err)
+
+	err = g.RunBlocking(context.Background())
```

Comment:
This error isn't returned, shouldn't this be returned to the caller?

Reply:
I addressed this as part of the refactoring to wait for documents to be stored in ES, in [`fd73107` (#14935)](https://github.com/elastic/apm-server/pull/14935/commits/fd731071062bc06790269e8492605fa665101098)